File: Leetcode/cpp/easy/242.cpp
/*
 * @lc app=leetcode id=242 lang=cpp
 *
 * [242] Valid Anagram
 */


// @lcpr-template-start
#include <bits/stdc++.h>
using namespace std;
// @lcpr-template-end
// @lc code=start
class Solution1 {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) return false;
        unordered_map<char, int> cnt_s, cnt_t;
        for (char c : s) cnt_s[c]++;
        for (char c : t) cnt_t[c]++;
        return cnt_s == cnt_t;
    }
};

class Solution2 {
public:
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) return false;
        unordered_map<char, int> cnt;
        for (char c : s) cnt[c]++;
        for (char c : t) cnt[c]--;
        return all_of(cnt.begin(), cnt.end(), [](const auto& p) { return p.second == 0; });
    }
};

// using Solution = Solution1;
using Solution = Solution2;
// @lc code=end

File: Leetcode/rust/easy/242.rs
/*
 * @lc app=leetcode id=242 lang=rust
 *
 * [242] Valid Anagram
 */


// @lcpr-template-start
struct Solution;
// @lcpr-template-end
// @lc code=start
use std::collections::HashMap;

struct Solution1;
impl Solution1 {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() { return false; }
        let mut cnt_s = HashMap::new();
        let mut cnt_t = HashMap::new();
        for c in s.chars() {
            *cnt_s.entry(c).or_insert(0) += 1;
        }
        for c in t.chars() {
            *cnt_t.entry(c).or_insert(0) += 1;
        }
        cnt_s == cnt_t
    }
}

struct Solution2;
impl Solution2 {
    pub fn is_anagram(s: String, t: String) -> bool {
        if s.len() != t.len() { return false; }
        let mut cnt = HashMap::new();
        for c in s.chars() {
            *cnt.entry(c).or_insert(0) += 1;
        }
        for c in t.chars() {
            *cnt.entry(c).or_insert(0) -= 1;
        }
        cnt.values().all(|&v| v == 0)
    }
}

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        // Solution1::is_anagram(s, t)
        Solution2::is_anagram(s, t)
    }
}
// @lc code=end

